File: packages/snow_draw_core/lib/draw/utils/binary_heap.dart
/// A generic min-heap that orders elements by a caller-supplied score.
///
/// Used by the elbow A* grid router and available for reuse by other
/// subsystems that need priority-queue behavior.
class BinaryHeap<T> {
  /// Creates a heap that orders elements by the given scoring function
  /// (lowest first).
  BinaryHeap(this._score);

  final double Function(T) _score;
  final List<T> _content = [];

  /// Whether the heap contains no elements.
  bool get isEmpty => _content.isEmpty;

  /// Whether the heap contains at least one element.
  bool get isNotEmpty => _content.isNotEmpty;

  /// Adds [element] to the heap.
  void push(T element) {
    _content.add(element);
    _sinkDown(_content.length - 1);
  }

  /// Removes and returns the element with the lowest score.
  ///
  /// Returns `null` if the heap is empty.
  T? pop() {
    if (_content.isEmpty) {
      return null;
    }
    final result = _content.first;
    final end = _content.removeLast();
    if (_content.isNotEmpty) {
      _content[0] = end;
      _bubbleUp(0);
    }
    return result;
  }

  /// Whether the heap contains [element].
  bool contains(T element) => _content.contains(element);

  /// Re-positions [element] after its score has changed.
  ///
  /// If [element] is not in the heap this is a no-op.
  void rescore(T element) {
    final index = _content.indexOf(element);
    if (index >= 0) {
      _sinkDown(index);
    }
  }

  void _sinkDown(int n) {
    final element = _content[n];
    final elementScore = _score(element);
    while (n > 0) {
      final parentN = ((n + 1) >> 1) - 1;
      final parent = _content[parentN];
      if (elementScore < _score(parent)) {
        _content[parentN] = element;
        _content[n] = parent;
        n = parentN;
      } else {
        break;
      }
    }
  }

  void _bubbleUp(int n) {
    final length = _content.length;
    final element = _content[n];
    final elemScore = _score(element);

    while (true) {
      final child2N = (n + 1) << 1;
      final child1N = child2N - 1;
      int? swap;
      var child1Score = 0.0;

      if (child1N < length) {
        final child1 = _content[child1N];
        child1Score = _score(child1);
        if (child1Score < elemScore) {
          swap = child1N;
        }
      }

      if (child2N < length) {
        final child2 = _content[child2N];
        final child2Score = _score(child2);
        if (child2Score < (swap == null ? elemScore : child1Score)) {
          swap = child2N;
        }
      }

      if (swap != null) {
        _content[n] = _content[swap];
        _content[swap] = element;
        n = swap;
      } else {
        break;
      }
    }
  }
}

File: packages/snow_draw_rust_engine/src/draw/utils/binary_heap.rs
#![allow(dead_code)]

/// A generic min-heap that orders elements by a caller-supplied score.
///
/// Used by the elbow A* grid router and reusable by other subsystems that
/// need priority-queue behavior.
pub struct BinaryHeap<T, F>
where
    F: Fn(&T) -> f64,
{
    score: F,
    content: Vec<T>,
}

impl<T, F> BinaryHeap<T, F>
where
    F: Fn(&T) -> f64,
{
    /// Creates a heap that orders elements by the given scoring function
    /// (lowest first).
    pub fn new(score: F) -> Self {
        Self {
            score,
            content: Vec::new(),
        }
    }

    /// Whether the heap contains no elements.
    pub fn is_empty(&self) -> bool {
        self.content.is_empty()
    }

    /// Whether the heap contains at least one element.
    pub fn is_not_empty(&self) -> bool {
        !self.content.is_empty()
    }

    /// Adds `element` to the heap.
    pub fn push(&mut self, element: T) {
        self.content.push(element);
        let last_index = self.content.len() - 1;
        self.sink_down(last_index);
    }

    /// Removes and returns the element with the lowest score.
    ///
    /// Returns `None` if the heap is empty.
    pub fn pop(&mut self) -> Option<T> {
        if self.content.is_empty() {
            return None;
        }

        let result = self.content.swap_remove(0);
        if !self.content.is_empty() {
            self.bubble_up(0);
        }
        Some(result)
    }

    fn score_at(&self, index: usize) -> f64 {
        (self.score)(&self.content[index])
    }

    fn sink_down(&mut self, mut index: usize) {
        while index > 0 {
            let parent_index = ((index + 1) >> 1) - 1;
            if self.score_at(index) < self.score_at(parent_index) {
                self.content.swap(index, parent_index);
                index = parent_index;
            } else {
                break;
            }
        }
    }

    fn bubble_up(&mut self, mut index: usize) {
        let length = self.content.len();
        loop {
            let child2_index = (index + 1) << 1;
            let child1_index = child2_index - 1;
            let mut swap_index: Option<usize> = None;

            if child1_index < length && self.score_at(child1_index) < self.score_at(index) {
                swap_index = Some(child1_index);
            }

            if child2_index < length {
                let compare_index = swap_index.unwrap_or(index);
                if self.score_at(child2_index) < self.score_at(compare_index) {
                    swap_index = Some(child2_index);
                }
            }

            if let Some(next_index) = swap_index {
                self.content.swap(index, next_index);
                index = next_index;
            } else {
                break;
            }
        }
    }
}

impl<T, F> BinaryHeap<T, F>
where
    T: PartialEq,
    F: Fn(&T) -> f64,
{
    /// Whether the heap contains `element`.
    pub fn contains(&self, element: &T) -> bool {
        self.content.contains(element)
    }

    /// Re-positions `element` after its score has changed.
    ///
    /// If `element` is not in the heap this is a no-op.
    pub fn rescore(&mut self, element: &T) {
        if let Some(index) = self.content.iter().position(|item| item == element) {
            self.sink_down(index);
        }
    }
}
